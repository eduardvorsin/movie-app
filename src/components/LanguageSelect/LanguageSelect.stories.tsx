import type { Meta, StoryObj } from '@storybook/react';
import LanguageSelect, { Props } from './LanguageSelect';
import { useTranslation } from 'react-i18next';

const meta: Meta<typeof LanguageSelect> = {
	title: 'components/LanguageSelect',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the button'
		},
		testId: {
			description: 'Id for testing the component',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
	},
	component: LanguageSelect,
	parameters: {
		docs: {
			description: {
				component: 'The Language Select component is designed to provide users with the ability to select the language of the application interface, providing localization and personalization of the user experience. It allows you to easily switch between available languages, improving usability and adapting content to the preferences of the end user.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof LanguageSelect>;

const LanguageSelectWithHooks = (props: Omit<Props, 'dictionary'>) => {
	const { t } = useTranslation('common');
	const dictionary = { label: t('languageSelect.label') };

	return (
		<LanguageSelect
			{...props}
			dictionary={dictionary}
			className='max-w-[200px]'
		/>);
}

export const Default: Story = {
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => <LanguageSelectWithHooks {...args} />
};
