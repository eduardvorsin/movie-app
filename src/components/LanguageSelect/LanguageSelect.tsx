'use client';

import { useTranslation } from '@/i18n/client';
import { Locales, fallbackLng } from '@/i18n/settings';
import { GeneralProps } from '@/types/shared';
import { useParams, usePathname, useRouter } from 'next/navigation';
import Select from '../Select/Select';
import { useId, useState } from 'react';

type LanguageOption = {
	label: 'RU' | 'EN',
	value: 'ru' | 'en',
}

const languageOptions: LanguageOption[] = [
	{
		label: 'RU',
		value: 'ru',
	},
	{
		label: 'EN',
		value: 'en',
	},
]

type Props = {
	dictionary: Record<'label', string>,
} & GeneralProps;

export default function LanguageSelect({
	className,
	testId,
	dictionary,
}: Props) {
	const id = useId();
	const currentLang = useParams()?.lang as Locales ?? fallbackLng;
	const { i18n } = useTranslation(currentLang);
	const [lang, setLang] = useState<Locales>(currentLang);
	const pathname = usePathname();
	const router = useRouter();

	const changeHandler = () => {
		const otherLang = lang === 'ru' ? 'en' : 'ru';
		setLang(otherLang);
		i18n.changeLanguage(otherLang);
		router.replace(pathname.replace(currentLang, otherLang));
	}

	return (
		<Select
			className={className}
			testId={testId}
			id={id}
			labelHidden
			label={dictionary.label}
			name='language select'
			options={languageOptions}
			value={lang}
			closeMenuOnSelect
			closeMenuOnScroll
			openMenuOnFocus
			onChange={changeHandler}
			minMenuHeight={69}
			maxMenuHeight={69}
		/>
	);
}