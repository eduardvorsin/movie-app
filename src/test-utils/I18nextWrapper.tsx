import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;