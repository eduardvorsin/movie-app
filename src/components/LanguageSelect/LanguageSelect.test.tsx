import { render, screen } from "@testing-library/react";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";
import LanguageSelect from "./LanguageSelect";

jest.mock('next/navigation', () => ({
	useParams: () => jest.fn(),
	useRouter: () => jest.fn(),
	usePathname: () => jest.fn(),
}));

const dictionary = { label: 'Language selection' };

describe('LanguageSelect tests', () => {
	it('is rendered correctly', async () => {
		render(
			<LanguageSelect
				dictionary={dictionary}
				testId='test-language-select'
			/>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-language-select')).toBeInTheDocument();
	});

	it('is a basic snapshot', async () => {
		render(
			<LanguageSelect
				dictionary={dictionary}
				testId='test-language-select'
			/>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-language-select')).toMatchSnapshot();
	});
});

describe('LanguageSelect integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('languageSelect.label'));
		const dictionary = { label: i18next.t('languageSelect.label') };
		render(
			<LanguageSelect
				dictionary={dictionary}
				testId='test-language-select'
			/>,
			{ wrapper: I18nextWrapper }
		);
		expect(await screen.findByText<HTMLLabelElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('languageSelect.label'));
		const dictionary = { label: i18next.t('languageSelect.label') };
		render(
			<LanguageSelect
				dictionary={dictionary}
				testId='test-language-select'
			/>,
			{ wrapper: I18nextWrapper }
		);
		expect(await screen.findByText<HTMLLabelElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');
		const dictionary = { label: i18next.t('languageSelect.label') };
		render(
			<LanguageSelect
				dictionary={dictionary}
				testId='test-language-select'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-language-select')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');
		const dictionary = { label: i18next.t('languageSelect.label') };
		render(
			<LanguageSelect
				dictionary={dictionary}
				testId='test-language-select'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-language-select')).toMatchSnapshot();
	});
});