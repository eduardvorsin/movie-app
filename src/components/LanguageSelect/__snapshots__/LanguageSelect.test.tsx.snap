// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelect integration tests is a snapshot with English localization 1`] = `
<div
  class="w-full  "
  data-testid="test-language-select"
  id=":r9:"
  name="language select"
>
  <label
    class="block mb-1 sr-only"
    id=":ra:-label"
  >
    Language selection
  </label>
  <div
    class="relative"
  >
    <button
      aria-controls=":ra:-listbox"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby=":ra:-label"
      class="w-full text-left rounded-1 border-2  px-[0.375rem] py-[0.125rem] min-h-[2.5rem] text-100 text-neutral-700 dark:text-dark-neutral-700 bg-neutral-0 dark:bg-dark-neutral-200 enabled:hover:bg-neutral-100 dark:enabled:hover:bg-dark-neutral-250 focus:border-blue-500 dark:focus:border-blue-300 transition-colors duration-150 leading-2 truncate  border-neutral-300 dark:border-dark-neutral-350 enabled:hover:border-neutral-400 dark:enabled:hover:border-dark-neutral-500 "
      role="combobox"
    >
      EN
    </button>
    <ul
      aria-labelledby=":ra:-label"
      class="w-full absolute z-300 mt-2 rounded-1 border-2 border-neutral-300 dark:border-dark-neutral-350 shadow-md bg-neutral-0 dark:bg-dark-neutral-250 overflow-y-auto transition-colors duration-150 hidden"
      id=":ra:-listbox"
      role="listbox"
      style="max-height: 69px; min-height: 69px;"
    >
      <li
        aria-selected="false"
        class="outline-none break-words px-[0.625rem] py-[0.375rem] text-100 leading-2 border-neutral-300 dark:border-dark-neutral-400 border-b-1 last:border-none transition-colors duration-150 select-none relative before:w-[0.125rem] before:h-full before:absolute before:left-0 before:top-0 before:transition-colors before:duration-150   bg-neutral-0 dark:bg-dark-neutral-250 text-neutral-800 dark:text-dark-neutral-800 "
        data-label="RU"
        data-value="ru"
        id=":ra:-list-option-RU"
        role="option"
        tabindex="-1"
      >
        RU
      </li>
      <li
        aria-selected="true"
        class="outline-none break-words px-[0.625rem] py-[0.375rem] text-100 leading-2 border-neutral-300 dark:border-dark-neutral-400 border-b-1 last:border-none transition-colors duration-150 select-none relative before:w-[0.125rem] before:h-full before:absolute before:left-0 before:top-0 before:transition-colors before:duration-150 bg-blue-200 text-blue-900 dark:text-blue-300 dark:bg-blue-900   "
        data-label="EN"
        data-value="en"
        id=":ra:-list-option-EN"
        role="option"
        tabindex="-1"
      >
        EN
      </li>
    </ul>
  </div>
</div>
`;

exports[`LanguageSelect integration tests is a snapshot with Russian localization 1`] = `
<div
  class="w-full  "
  data-testid="test-language-select"
  id=":rb:"
  name="language select"
>
  <label
    class="block mb-1 sr-only"
    id=":rc:-label"
  >
    Выбор языка
  </label>
  <div
    class="relative"
  >
    <button
      aria-controls=":rc:-listbox"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby=":rc:-label"
      class="w-full text-left rounded-1 border-2  px-[0.375rem] py-[0.125rem] min-h-[2.5rem] text-100 text-neutral-700 dark:text-dark-neutral-700 bg-neutral-0 dark:bg-dark-neutral-200 enabled:hover:bg-neutral-100 dark:enabled:hover:bg-dark-neutral-250 focus:border-blue-500 dark:focus:border-blue-300 transition-colors duration-150 leading-2 truncate  border-neutral-300 dark:border-dark-neutral-350 enabled:hover:border-neutral-400 dark:enabled:hover:border-dark-neutral-500 "
      role="combobox"
    >
      EN
    </button>
    <ul
      aria-labelledby=":rc:-label"
      class="w-full absolute z-300 mt-2 rounded-1 border-2 border-neutral-300 dark:border-dark-neutral-350 shadow-md bg-neutral-0 dark:bg-dark-neutral-250 overflow-y-auto transition-colors duration-150 hidden"
      id=":rc:-listbox"
      role="listbox"
      style="max-height: 69px; min-height: 69px;"
    >
      <li
        aria-selected="false"
        class="outline-none break-words px-[0.625rem] py-[0.375rem] text-100 leading-2 border-neutral-300 dark:border-dark-neutral-400 border-b-1 last:border-none transition-colors duration-150 select-none relative before:w-[0.125rem] before:h-full before:absolute before:left-0 before:top-0 before:transition-colors before:duration-150   bg-neutral-0 dark:bg-dark-neutral-250 text-neutral-800 dark:text-dark-neutral-800 "
        data-label="RU"
        data-value="ru"
        id=":rc:-list-option-RU"
        role="option"
        tabindex="-1"
      >
        RU
      </li>
      <li
        aria-selected="true"
        class="outline-none break-words px-[0.625rem] py-[0.375rem] text-100 leading-2 border-neutral-300 dark:border-dark-neutral-400 border-b-1 last:border-none transition-colors duration-150 select-none relative before:w-[0.125rem] before:h-full before:absolute before:left-0 before:top-0 before:transition-colors before:duration-150 bg-blue-200 text-blue-900 dark:text-blue-300 dark:bg-blue-900   "
        data-label="EN"
        data-value="en"
        id=":rc:-list-option-EN"
        role="option"
        tabindex="-1"
      >
        EN
      </li>
    </ul>
  </div>
</div>
`;

exports[`LanguageSelect tests is a basic snapshot 1`] = `
<div
  class="w-full  "
  data-testid="test-language-select"
  id=":r3:"
  name="language select"
>
  <label
    class="block mb-1 sr-only"
    id=":r4:-label"
  >
    Language selection
  </label>
  <div
    class="relative"
  >
    <button
      aria-controls=":r4:-listbox"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby=":r4:-label"
      class="w-full text-left rounded-1 border-2  px-[0.375rem] py-[0.125rem] min-h-[2.5rem] text-100 text-neutral-700 dark:text-dark-neutral-700 bg-neutral-0 dark:bg-dark-neutral-200 enabled:hover:bg-neutral-100 dark:enabled:hover:bg-dark-neutral-250 focus:border-blue-500 dark:focus:border-blue-300 transition-colors duration-150 leading-2 truncate  border-neutral-300 dark:border-dark-neutral-350 enabled:hover:border-neutral-400 dark:enabled:hover:border-dark-neutral-500 "
      role="combobox"
    >
      EN
    </button>
    <ul
      aria-labelledby=":r4:-label"
      class="w-full absolute z-300 mt-2 rounded-1 border-2 border-neutral-300 dark:border-dark-neutral-350 shadow-md bg-neutral-0 dark:bg-dark-neutral-250 overflow-y-auto transition-colors duration-150 hidden"
      id=":r4:-listbox"
      role="listbox"
      style="max-height: 69px; min-height: 69px;"
    >
      <li
        aria-selected="false"
        class="outline-none break-words px-[0.625rem] py-[0.375rem] text-100 leading-2 border-neutral-300 dark:border-dark-neutral-400 border-b-1 last:border-none transition-colors duration-150 select-none relative before:w-[0.125rem] before:h-full before:absolute before:left-0 before:top-0 before:transition-colors before:duration-150   bg-neutral-0 dark:bg-dark-neutral-250 text-neutral-800 dark:text-dark-neutral-800 "
        data-label="RU"
        data-value="ru"
        id=":r4:-list-option-RU"
        role="option"
        tabindex="-1"
      >
        RU
      </li>
      <li
        aria-selected="true"
        class="outline-none break-words px-[0.625rem] py-[0.375rem] text-100 leading-2 border-neutral-300 dark:border-dark-neutral-400 border-b-1 last:border-none transition-colors duration-150 select-none relative before:w-[0.125rem] before:h-full before:absolute before:left-0 before:top-0 before:transition-colors before:duration-150 bg-blue-200 text-blue-900 dark:text-blue-300 dark:bg-blue-900   "
        data-label="EN"
        data-value="en"
        id=":r4:-list-option-EN"
        role="option"
        tabindex="-1"
      >
        EN
      </li>
    </ul>
  </div>
</div>
`;
